frontend/rust-lib/flowy-document/src/event_handler.rs

```
@@ -401,3 +406,125 @@ pub(crate) async fn convert_data_to_json_handler(
+    })
+    .await?;
+
+  // let the upload happen in the background
+  let url_clone = url.clone();
+  tokio::spawn(async move {
+    if let Err(e) = file_service
+      .put_object(
+        url_clone,
+        ObjectValue {
+          raw: content.into(),
+          mime,
+        },
+      )
+      .await
+    {
+      tracing::error!("upload file failed: {}", e);
+    }
+  });
+
+  let local_file_path = params.local_file_path.to_owned();
+  Ok(AFPluginData(UploadedFilePB {
+    url,
+    local_file_path,
+  }))
+}
+
+pub(crate) async fn download_file_handler(
+  params: AFPluginData<UploadedFilePB>,
+  manager: AFPluginState<Weak<DocumentManager>>,
+) -> FlowyResult<()> {
+  let path = params.local_file_path.clone();
+
+  // if file already exist in user local disk, return
+  if tokio::fs::metadata(&path).await.is_ok() {
+    return Ok(());
+  }
+
+  let manager = upgrade_document(manager)?;
+  let file_service = manager
+    .get_file_storage_service()
+    .upgrade()
+    .ok_or(FlowyError::internal().with_context("The file storage service is already dropped"))?;
+
+  let object_value = file_service.get_object(params.url.clone()).await?;
+
+  // create file if not exist
+  let mut file = tokio::fs::OpenOptions::new()
+    .create(true)
+    .write(true)
+    .open(&path)
+    .await?;
+
+  let n = file.write(&object_value.raw).await?;
+  info!("downloaded {} bytes to file: {}", n, path);
+  Ok(())
+}
+
+// Handler for creating a new document
+pub(crate) async fn delete_uploaded_file_handler(
+  params: AFPluginData<UploadedFilePB>,
+  manager: AFPluginState<Weak<DocumentManager>>,
+) -> FlowyResult<()> {
+  let manager = upgrade_document(manager)?;
+  let file_service = manager
+    .get_file_storage_service()
+    .upgrade()
+    .ok_or(FlowyError::internal().with_context("The file storage service is already dropped"))?;
+  file_service.delete_object(params.url.clone()).await?;
+
+  tokio::fs::remove_file(&params.local_file_path).await?;
```

I think the frontend might can't handle the error when remove_file fail. So it doesn't need to propagate the remove_file error. We can just log it. But we can have a junk list somewhere. When user try to clean up the disk space. we can try to delete these files again.

---

I added a TODO for this. I suppose the frontend needs to retry the deletion operation when it can. not particular sure when it should trigger but mechanisms can be further discussed.

---

Yep. We can do more discussion later